hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+        val outputShapePType = outputPType.representation.fieldType("shape").asInstanceOf[PTuple]
+
+        val shapeSrvb = new StagedRegionValueBuilder(mb, outputShapePType)
+
+        val childShapeTuple = new CodePTuple(childPType.shape.pType, region, childEmitter.outputShape)
+
+        val reindexShape = indexExpr.map {childIndex =>
+          if (childIndex < childPType.nDims) {
+            Code(
+              shapeSrvb.addLong(childShapeTuple(childIndex)),
+              shapeSrvb.advance()
+            )
+          }
+          else {
+            Code(
+              shapeSrvb.addLong(1L),
+              shapeSrvb.advance()
+            )
+          }
+        }
+
+        val setup = Code(childEmitter.setup, shapeSrvb.start(), reindexShape)
+
+        new NDArrayEmitter(mb, indexExpr.length, shapeSrvb.end(), outputShapePType, outputPType.elementType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+            val concreteIdxsForChild = Array.tabulate(childEmitter.nDims) { childDim =>
+              val parentDim = indexExpr.indexOf(childDim)
+              idxVars(parentDim)
+            }
+            childEmitter.outputElement(concreteIdxsForChild)
+          }
+        }
+
       case _ =>
         val ndt = emit(x, env, er, None)
         val setup = Code(ndt.setup)
         val xP = x.pType.asInstanceOf[PNDArray]
 
         new NDArrayEmitter(mb, nDims, xP.representation.loadField(er.region, ndt.value[Long], "shape"),
           xP.representation.fieldType("shape").asInstanceOf[PTuple], xP.elementType, setup) {
-          override def outputElement(idxVars: Seq[ClassFieldRef[Long]]): Code[_] = {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
             val elementLocation = xP.getElementPosition(idxVars, ndt.value[Long], er.region, mb)
             Region.loadIRIntermediate(outputElementPType)(elementLocation)
           }
         }
     }
   }
 }
+
+object NDArrayEmitter {
+
+  def zeroBroadcastedDims2(mb: MethodBuilder, loopVars: Array[Code[Long]], nDims: Int, shape: Code[Long], shapePType: PTuple): Array[Code[Long]] = {
+    val broadcasted = 0L
+    val notBroadcasted = 1L
+    def getShapeAtIdx(i: Int): Code[Long] = Region.loadLong(shapePType.loadField(shape, i))
+
+    Array.tabulate(nDims)(dim => (getShapeAtIdx(dim) > 1L).mux(notBroadcasted, broadcasted) * loopVars(dim))
+  }
+
+  def unifyShapes(mb: MethodBuilder, leftShape: Code[Long], leftShapePType: PTuple, rightShape: Code[Long],
+    rightShapePType: PTuple, region: Code[Region]): (Code[_], Code[Long], PTuple) = {
+    assert(leftShapePType == rightShapePType)
+
+    val unifiedPType = leftShapePType
+    def getShapeAtIdx(shape: Code[Long], idx: Int): Code[Long] = Region.loadLong(unifiedPType.loadField(shape, idx))
+    val unifiedShapeAddress = mb.newField[Long]
+
+    val setupList = Array.tabulate(unifiedPType.size) { idx =>
+      val left = getShapeAtIdx(leftShape, idx)
```

Comment:
So, this is going to allocate a shape tuple for each dimension you look up. Which makes this calculation (which should be constant time in the number of dimensions and number of IR nodes) `O(number_of_dimensions^number_of_ir_nodes)`

Reply:
We need to make the changes I mentioned earlier about `NDArrayEmitter`'s having `Array[Code[Long]]` I think that's the right way to fix all this. No allocations.